¡Hola Bruno!

Mi nombre es Facundo Lozano! Un gusto conocerte, seré tu revisor en este proyecto.

Cuando enccuentro un error por primera vez, simplemente lo señalaré, te dejaré encontrarlo y arreglarlo tú cuenta. Además, a lo largo del texto iré haciendo algunas observaciones sobre mejora en tu código y también haré comentarios sobre tus percepciones sobre el tema. Pero si aún no puedes realizar esta tarea, te daré una pista más precisa en la próxima iteración y también algunos ejemplos prácticos. Estaré abierto a comentarios y discusiones sobre el tema.

Encontrará mis comentarios a continuación: **no los mueva, modifique ni elimine**.

Puedes encontrar mis comentarios en cuadros verdes, amarillos o rojos como este:

<div class="alert alert-block alert-success">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Exito. Todo se ha hecho de forma exitosa.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Observación. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">

<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Necesita arreglos. Este apartado necesita algunas correcciones. El trabajo no puede ser aceptado con comentarios rojos. 
</div>

Puedes responder utilizando esto:

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante.</b> <a class="tocSkip"></a>
</div>

# Déjame escuchar la música

<div class="alert alert-block alert-success">
<b>Review General. (Iteración 2) </b> <a class="tocSkip"></a>

Hola Bruno! Felicitaciones porque has corregido los detalles marcados en nuestra iteración anterior! Ahora si este proyecto está en total condiciones de ser aprobado, bien hecho!
    
Éxitos en tu camino dentro del mundo de los datos y saludos!

<div class="alert alert-block alert-success">
<b>Review General. (Iteración 1) </b> <a class="tocSkip"></a>

Bruno, quería tomarme este tiempo al inicio de tu proyecto para comentarte mis apreciaciones generales de esta iteración de tu entrega.

Siempre me gusta comenzar dando la bienvenida al mundo de los datos a los estudiantes, te deseo lo mejor y espero que consigas lograr tus objetivos. Personalmente me gusta brindar el siguiente consejo, "Está bien equivocarse, es normal y es lo mejor que te puede pasar. Aprendemos de los errores y eso te hará mejor programador ya que podrás descubrir cosas a medida que avances y son estas cosas las que te darán esa experiencia para ser un gran cientifico de datos."
    
Ahora si yendo a esta notebook. Te quiero felicitar porque has tenido un gran desempeño a lo largo del proyecto, te has destacado por tu gran capacidad de lógica y de manejo de python, muy bien hecho! Sin embargo, hemos tenido un pequeño detalle que debemos corregir en una de las respuestas a nuestras observaciones del inicio del proyecto, te he dejado un comentario dando un contexto del porque del error.

Recuerda que estoy aquí para ayudarte y si tienes alguna duda puedes preguntarme que con gusto te ayudaré.

Saludos y te espero con ansias en nuestra próxima iteración!

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [4]:
import pandas as pd# Importar pandas


Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [5]:
df=pd.read_csv('/datasets/music_project_en.csv')
# Leer el archivo y almacenarlo en df


Muestra las 10 primeras filas de la tabla:

In [6]:
print(df.head(10))# Obtener las 10 primeras filas de la tabla df


     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [7]:
print(df.info())# Obtener la información general sobre nuestros datos


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Bien hecho Bruno, muy bien aplicados los métodos para obtener información! <div>

Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `Detecta el tercer problema por tu cuenta y descríbelo aquí`.La palabra 'userID' debería estár separada por un '_' para no confundir palabras cuando se realice la limpieza.




<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Excelente Bruno! Muy buen descubrimiento, bien hecho!
    
<div>

### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`
    En las filas, tenemos datos representados como objetos, lo que generalmente indica que contienen cadenas de texto u otros tipos de datos mixtos. Esto significa que las entradas en las filas pueden ser identificadores de usuario, títulos de canciones, nombres de artistas, géneros de música, ciudades, horarios y días de la semana. Para entender lo que almacenan las columnas, podemos examinar los nombres de las columnas y sus descripciones. 
       
`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`
    Para determinar si hay suficientes datos para proporcionar respuestas a una hipótesis, necesitamos evaluar la calidad y la cantidad de los datos disponibles. Según la información proporcionada por df.info(), podemos observar que hay 65079 entradas en total en el DataFrame. Sin embargo, también notamos que algunas columnas tienen valores faltantes (Non-Null Count es menor que el número total de registros), lo que podría afectar la capacidad de proporcionar respuestas precisas. Por lo tanto, es posible que necesitemos más información o realizar más análisis para confirmar nuestras hipótesis.
       

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`
    Sí, al observar la información proporcionada por df.info(), notamos que algunas columnas tienen valores faltantes (Non-Null Count es menor que el número total de registros). Esto indica la presencia de valores ausentes en esas columnas. Además, todas las columnas tienen el tipo de datos 'object', lo que sugiere que podrían contener una variedad de tipos de datos, incluidas las cadenas de texto. Si bien no se menciona explícitamente en df.info(), también podríamos sospechar la presencia de otros problemas en los datos, como valores duplicados, que no se revelarían en esta información. Es importante realizar un análisis más detallado para abordar estos problemas y garantizar la calidad de los datos.

       

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 2)</b> <a class="tocSkip"></a>

Excelente Bruno! Muy bien hecho, buenos descubrimientos! <div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Aquí podríamos hacer uso de la información que nos proporcionó el método info(), te invito a observar nuevamente esto y poder dar una nueva respuesta.    
<div>

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [8]:
print(df.columns)# Muestra los nombres de las columnas


Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Muy bien! Has aplicado correctamente el método .columns </div>

Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [9]:
for i in df.columns:
    df.rename(columns={i: i.lower()}, inplace=True)# Bucle en los encabezados poniendo todo en minúsculas


In [10]:
df.columns

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Excelente Bruno! Muy buen procedimiento y buen método para renombrar las columnas de nuestro dataframe.
    

Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [11]:
for i in df.columns:
    df.rename(columns={i: i.strip()}, inplace=True)# Bucle en los encabezados eliminando los espacios


In [12]:
df.columns

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Nuevamente muy bien! Hemos agregado el método .strip() para eliminar los espacios. Bien hecho! 

Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [13]:
df.rename(columns={'userid': 'user_id'}, inplace=True)# Cambiar el nombre de la columna "userid"


In [14]:
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Una utilización perfecta del método .rename(), bien resuelto Bruno! Bien hecho!

Comprueba el resultado. Muestra los encabezados una vez más:

In [15]:
print(df.columns)# Comprobar el resultado: la lista de encabezados


Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [16]:
print(df.isna().sum())# Calcular el número de valores ausentes


user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Muy buena implementación para resolver la consigna. Sigue así!
    

No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [17]:
columns_to_replace=['track', 'artist', 'genre']
df[columns_to_replace] = df[columns_to_replace].fillna('unknown')# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Felicitaciones! Implementaste una lógica de gran utilidad para evitar repeteciones en tu código y has completado con el valor 'unknown' como se ha solicitado, perfecto!.</div>

Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [18]:
print(df.isna().sum())# Contar valores ausentes


user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [19]:
print(df.duplicated().sum())# Contar duplicados explícitos


3826


Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [20]:
df=df.drop_duplicates()# Eliminar duplicados explícitos


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Muy bien, Data Scientist nato! Excelente implementación para corroborar duplicados y eliminarlos con drop_diplicates(). Recuerda que siempre es una buena practica aplicar el método **reset_index()** luego de eliminar valores. Sigue así!</div>

Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [21]:
print(df.duplicated().sum())# Comprobar de nuevo si hay duplicados



0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [22]:
column_genre=df['genre']
unique_values=column_genre.unique()
print(unique_values)# Inspeccionar los nombres de géneros únicos


['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hip' 'jazz' 'postrock' 'latin'
 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental' 'rusrock'
 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'hiphop' 'drum' 'extrememetal' 'türkçe'
 'experimental' 'easy' 'metalcore' 'modern' 'argentinetango' 'old' 'swing'
 'breaks' 'eurofolk' 

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Buena implemetación! Bien hecho, a este paso le podríamos agregar el método .sorted() que nos sería de gran utilidad ya que nos permitiría ver con un orden alfabético dichas columnas. Imagina que hubo un error y tenemos dos valores "garage" y "garagee", son distintos aunque nosotros sabemos que son lo mismo. Investiga .sorted()</div>


Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [23]:
def replace_wrong_genres(wrong_genres,correct_genres):
    for i in wrong_genres:
        df['genre']=df['genre'].replace(i, correct_genres)
    return df
    # Función para reemplazar duplicados implícitos


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Muy bien hecho! Buena implementación lógica, has resuelto correctamente lo solicitado por la consigna!<div>
    

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [25]:
wrong=['hip','hop','hip-hop']
new=('hiphop')

replace_wrong_genres(wrong,new)# Eliminar duplicados implícitos


user_id                              track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
65074  729CBB09                            My Name            McLean   
65075  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
65076  C5E3A0D5                          Jalopiina           unknown   
65077  321D0506                      Freight Train     Chas McDevitt   
65078  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfield  14:07:09     Friday  
2             pop  Shelbyville  20:58:07  Wednesday  
3            folk  Shelbyville  08:37:09     Monday  
4           dance  Springfield  08:34:34     Monday  
...           ...          ...       ...        ...  
65074         rnb  Springfield  13:32:28  Wednesday  
65075      hiphop  Shelbyville  10:00:00     Monday  
65076  industrial  Springfield  20:09:26     Friday  
65077        rock  Springfield  21:43:59     Friday  
65078     country  Springfield  21:59:46     Friday  

[61253 rows x 7 columns]

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [26]:
print(column_genre.unique())# Comprobación de duplicados implícitos


['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hiphop' 'jazz' 'postrock'
 'latin' 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental'
 'rusrock' 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'drum' 'extrememetal' 'türkçe' 'experimental' 'easy'
 'metalcore' 'modern' 'argentinetango' 'old' 'swing' 'breaks' 'eurofolk'
 'stone

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Nuevamente excelente! Has aplicado la función creada y se reemplazaron correctamente los valores. Recuerda la posibilidad de visualizar esta lista con el método **.sorted()**
    
<div>

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.`

Al analizar los duplicados en los datos, observé que había algunos valores que representaban la misma fila, esto con los valores explícitos para el caso de los implícitos considero que se requiere más expertís en el tema a tratar. Por ejemplo con el género de música que se encontraba escrito de manera ligeramente diferente, como 'hip', 'hop' y 'hip-hop'. Esto podría ser problemático si queremos analizar o visualizar los datos de manera precisa, ya que estos valores deberían tratarse como uno solo.

Para abordar esta situación, utilicé la función replace_wrong_genres() para reemplazar los valores incorrectos ('hip', 'hop', 'hip-hop') por el valor correcto ('hiphop'). Esto ayudó a consolidar los diferentes nombres de géneros en uno solo, lo que facilita el análisis y la interpretación de los datos.

Después de aplicar esta corrección, obtuve un conjunto de datos limpio y consistente en el que los géneros de música están representados de manera uniforme. Esto me permite realizar análisis más precisos y obtener insights más confiables a partir de los datos.


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Excelente! A medida que continúes con tus estudios aprenderás técnicas para como manejar estos casos. Muy bien hecho Bruno!</div>

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [21]:
canciones_reproducidas_por_ciudad = df.groupby(by='city')['track'].count()

print(canciones_reproducidas_por_ciudad)# Contar las canciones reproducidas en cada ciudad


city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Muy bien hecho Bruno, has aplicado la agrupación correcta con el método correcto

</div>

`Comenta tus observaciones aquí` De primera instancia podemos notar que el número de canciones reproducidas en Springfield es de poco más del doble que Shelbyville

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [27]:
canciones_por_dia = df.groupby('day')['track'].count()
dias_interesantes = ['Monday', 'Wednesday', 'Friday']
canciones_reproducidas = canciones_por_dia.loc[dias_interesantes]

print(canciones_reproducidas)# Calcular las canciones reproducidas en cada uno de los tres días


day
Monday       21354
Wednesday    18059
Friday       21840
Name: track, dtype: int64


<div class="alert alert-block alert-warning">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Aquí lo hemos resuelto bien nuevamente, podríamos haber implementado el mismo método que el anterior pero cambiando la agrupación, trata de modificarlo con un groupby() para que sea más legible y conciso.
</div>

`Comenta tus observaciones aquí`El día con mayores reproducciones fue el viernes y el de menor número el miércoles.

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [24]:
def number_tracks(df, day, city):# Declara la función number_tracks() con dos parámetros: day= y city=.

    df_day = df[df['day'] == day]# Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=

    df_filtered = df_day[df_day['city'] == city]# Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=

    numero_canciones = df_filtered['user_id'].count()# Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()

    return numero_canciones# Devolve el número de valores de la columna 'user_id'

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Excelente forma de resolverlo! Realizaste 2 filtros en una única sequencia que te brindarían el resultado posteriormente al utilizar .count(), bien hecho!.</div>

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [25]:
print(number_tracks(df,'Monday','Springfield'))# El número de canciones reproducidas en Springfield el lunes


15740


In [26]:
print(number_tracks(df,'Monday','Shelbyville'))# El número de canciones reproducidas en Shelbyville el lunes


5614


In [28]:
print(number_tracks(df,'Wednesday','Springfield'))# El número de canciones reproducidas en Springfield el miércoles


11056


In [29]:
print(number_tracks(df,'Wednesday','Shelbyville'))# El número de canciones reproducidas en Shelbyville el miércoles


7003


In [30]:
print(number_tracks(df,'Friday','Springfield'))# El número de canciones reproducidas en Springfield el viernes


15945


In [31]:
print(number_tracks(df,'Friday','Shelbyville'))# El número de canciones reproducidas en Shelbyville el viernes


5895


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Muy bien estas últimas implementaciones!

**Conclusiones**

`Comenta si la hipótesis es correcta o se debe rechazar. Explica tu razonamiento.` Citando de primera instancia a la hipótesis "La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes." Definitivamente existe una diferencia en el consumo de música entre ambas ciudades, sin embargo, entre el día lunes y el día viernes la diferencia de consumo de música para ambas ciudades es muy similar, por lo cual al menos en esta situación no afirmaría que hay diferencia entre estos días.


[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

`Resume aquí tus conclusiones sobre la hipótesis.`

Conclusiones:
Dado que la hipóstesis afirma que existe diferencia entre los días y las ciudades en cuanto al comportamiento de consumo de música, para el caso de las ciudades si se concluye que hay una diferencia en cuanto al consumo dando valores muy superiores de Springfield (Aproximadamente del triple para el caso de análisis de Viernes y Lunes, y prácticamente del doble para el Miércoles). Para el caso específico de los días, el consumo de canciones en días Lunes y Viernes no estuvo muy separado del todo, lo cual a pesar de que hay una diferencia, se podría considerar mínima.

Se podría hacer un análisis más a fondo con otros días y tomar en cuenta la densidad poblacional o número de habitantes para poder brindar datos que permitan comparar el volumen de consumo 


<div class="alert alert-block alert-success">
<b>Comentario final. (Iteración 1) </b> <a class="tocSkip"></a>

Felicitaciones Bruno, hemos llegado al final de la notebook y tu desempeño fue increíble. Estas conclusiones reflejan y demuestran tu compromiso con el trabajo y lo valoramos, tanto tus aplicaciones de código como tus interpretaciones demuestran tu comprensión de lo trabajado. Felicitaciones! 

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)